### ISSUE TITLE
Envoy Cluster Wide Config cannot resolve OAuth2 filter

### ISSUE THREAD

**Issue by gsharma-jiggzy** [cilium/cilium#24848]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I was try to install EnvoyClusterWideConfig to require auth via envoy and got found an error in the cilium-agent logs

` unable to resolve \"type.googleapis.com/envoy.extensions.filters.http.oauth2.v3.OAuth2\": \"not found\"): {\n\t\"@type\": \"type.googleapis.com/envoy.config.listener.v3.Listener\`

This was generated by applying doing kubectl apply -f [clusterwideconfig.yaml](https://github.com/cilium/cilium/files/11215236/clusterwideconfig.yaml.txt)

and found the error by tailing one of the cilium pods and looking at the agent logs.

It looks like the auth filter pointed out by the cilium docs is not validating.

### Cilium Version

1.13.0

### Kernel Version

Linux node-worker-1 5.15.79-51.138.amzn2.x86_64 #1 SMP Mon Nov 28 18:44:27 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

EKS 1.25

### Relevant log output

```shell
Ignoring invalid CiliumEnvoyConfig JSON (proto: (line 1:406): unable to resolve \"type.googleapis.com/envoy.extensions.filters.http.oauth2.v3.OAuth2\": \"not found\"): {\n\t\"@type\": \"type.googleapis.com/envoy.config.listener.v3.Listener\",\n\t\"filterChains\": [\n\t\t{\n\t\t\t\"filterChain │
│ Match\": {\n\t\t\t\t\"transportProtocol\": \"raw_buffer\"\n\t\t\t},\n\t\t\t\"filters\": [\n\t\t\t\t{\n\t\t\t\t\t\"name\": \"envoy.filters.network.http_connection_manager\",\n\t\t\t\t\t\"typedConfig\": {\n\t\t\t\t\t\t\"@type\": \"type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionMa │
│ nager\",\n\t\t\t\t\t\t\"httpFilters\": [\n\t\t\t\t\t\t\t{\n\t\t\t\t\t\t\t\t\"name\": \"envoy.filters.http.oauth2\",\n\t\t\t\t\t\t\t\t\"typed_config\": {\n\t\t\t\t\t\t\t\t\t\"@type\": \"type.googleapis.com/envoy.extensions.filters.http.oauth2.v3.OAuth2\",\n\t\t\t\t\t\t\t\t\t\"config\": {\n\t\t\t\t\t\t\t\t\t\t\"auth_scopes\": │
│  [\n\t\t\t\t\t\t\t\t\t\t\t\"openid\"\n\t\t\t\t\t\t\t\t\t\t],\n\t\t\t\t\t\t\t\t\t\t\"authorization_endpoint\": \"https://keycloak.mydomain.com/auth/realms/myrealm/protocol/openid-connect/auth\",\n\t\t\t\t\t\t\t\t\t\t\"credentials\": {\n\t\t\t\t\t\t\t\t\t\t\t\"client_id\": \"envoy-clusterwide\",\n\t\t\t\t\t\t\t\t\t\t\t\"token_ │
│ secret\": {\n\t\t\t\t\t\t\t\t\t\t\t\t\"generic_secret\": {\n\t\t\t\t\t\t\t\t\t\t\t\t\t\"secret\": {\n\t\t\t\t\t\t\t\t\t\t\t\t\t\t\"inline_string\": \"SECRETS\"\n\t\t\t\t\t\t\t\t\t\t\t\t\t}\n\t\t\t\t\t\t\t\t\t\t\t\t},\n\t\t\t\t\t\t\t\t\t\t\t\t\"hmac_secret\": {\n\t\t\t\t\t\t\t\t\t\t\t\t\t\"generic_se │
│ cret\": {\n\t\t\t\t\t\t\t\t\t\t\t\t\t\t\"secret\": {\n\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\"inline_string\": \"redacted=\"\n\t\t\t\t\t\t\t\t\t\t\t\t\t\t}\n\t\t\t\t\t\t\t\t\t\t\t\t\t},\n\t\t\t\t\t\t\t\t\t\t\t\t\t\"name\": \"hmac\"\n\t\t\t\t\t\t\t\t\t\t\t\t},\n\t\t\t\t\t\t\t\t\t\t\t\t\"name\": \"t │
│ oken\"\n\t\t\t\t\t\t\t\t\t\t\t}\n\t\t\t\t\t\t\t\t\t\t},\n\t\t\t\t\t\t\t\t\t\t\"forward_bearer_token\": true,\n\t\t\t\t\t\t\t\t\t\t\"redirect_path_matcher\": {\n\t\t\t\t\t\t\t\t\t\t\t\"path\": {\n\t\t\t\t\t\t\t\t\t\t\t\t\"exact\": \"/oauth2/callback\"\n\t\t\t\t\t\t\t\t\t\t\t}\n\t\t\t\t\t\t\t\t\t\t},\n\t\t\t\t\t\t\t\t\t\t\"re │
│ direct_uri\": \"http://%REQ(:authority)%/oauth2/callback\",\n\t\t\t\t\t\t\t\t\t\t\"signout_path\": {\n\t\t\t\t\t\t\t\t\t\t\t\"path\": {\n\t\t\t\t\t\t\t\t\t\t\t\t\"exact\": \"/oauth2/signout\"\n\t\t\t\t\t\t\t\t\t\t\t}\n\t\t\t\t\t\t\t\t\t\t},\n\t\t\t\t\t\t\t\t\t\t\"token_endpoint\": {\n\t\t\t\t\t\t\t\t\t\t\t\"cluster\": \"oau │
│ th\",\n\t\t\t\t\t\t\t\t\t\t\t\"timeout\": \"5s\",\n\t\t\t\t\t\t\t\t\t\t\t\"uri\": \"https://keycloak.mydomain.com/auth/realms/myrealm/protocol/openid-connect/token\"\n\t\t\t\t\t\t\t\t\t\t}\n\t\t\t\t\t\t\t\t\t}\n\t\t\t\t\t\t\t\t}\n\t\t\t\t\t\t\t},\n\t\t\t\t\t\t\t{\n\t\t\t\t\t\t\t\t\"name\": \"envoy.filters.http.jwt_authn\",\n │
│ \t\t\t\t\t\t\t\t\"typed_config\": {\n\t\t\t\t\t\t\t\t\t\"@type\": \"type.googleapis.com/envoy.extensions.filters.http.jwt_authn.v3.JwtAuthentication\",\n\t\t\t\t\t\t\t\t\t\"providers\": {\n\t\t\t\t\t\t\t\t\t\t\"provider1\": {\n\t\t\t\t\t\t\t\t\t\t\t\"remote_jwks\": {\n\t\t\t\t\t\t\t\t\t\t\t\t\"cache_duration\": \"600s\",\n\ │
│ t\t\t\t\t\t\t\t\t\t\t\t\"http_uri\": {\n\t\t\t\t\t\t\t\t\t\t\t\t\t\"cluster\": \"oauth\",\n\t\t\t\t\t\t\t\t\t\t\t\t\t\"timeout\": \"5s\",\n\t\t\t\t\t\t\t\t\t\t\t\t\t\"uri\": \"https://keycloak.mydomain.com/auth/realms/myrealm/protocol/openid-connect/certs\"\n\t\t\t\t\t\t\t\t\t\t\t\t}\n\t\t\t\t\t\t\t\t\t\t\t}\n\t\t\t\t\t\t\t\ │
│ t\t\t}\n\t\t\t\t\t\t\t\t\t},\n\t\t\t\t\t\t\t\t\t\"rules\": [\n\t\t\t\t\t\t\t\t\t\t{\n\t\t\t\t\t\t\t\t\t\t\t\"match\": {\n\t\t\t\t\t\t\t\t\t\t\t\t\"prefix\": \"/\"\n\t\t\t\t\t\t\t\t\t\t\t},\n\t\t\t\t\t\t\t\t\t\t\t\"requires\": {\n\t\t\t\t\t\t\t\t\t\t\t\t\"provider_name\": \"provider1\"\n\t\t\t\t\t\t\t\t\t\t\t}\n\t\t\t\t\t\t\ │
│ t\t\t\t}\n\t\t\t\t\t\t\t\t\t]\n\t\t\t\t\t\t\t\t}\n\t\t\t\t\t\t\t},\n\t\t\t\t\t\t\t{\n\t\t\t\t\t\t\t\t\"name\": \"envoy.filters.http.router\",\n\t\t\t\t\t\t\t\t\"typed_config\": {}\n\t\t\t\t\t\t\t}\n\t\t\t\t\t\t]\n\t\t\t\t\t}\n\t\t\t\t}\n\t\t\t]\n\t\t}\n\t]\n}" subsys=k8s-apis-cilium.io-v2
```

### Anything else?

clusterwideconfig.yaml
---
apiVersion: cilium.io/v2
kind: CiliumClusterwideEnvoyConfig
metadata:
  name: envoy-oauth2-clusterwide
spec:
  resources:
    - '@type': type.googleapis.com/envoy.config.listener.v3.Listener
      filterChains:
        - filterChainMatch:
            transportProtocol: raw_buffer
          filters:
            - name: envoy.filters.network.http_connection_manager
              typedConfig:
                '@type': >-
                  type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
                httpFilters:
                  - name: envoy.filters.http.oauth2
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.oauth2.v3.OAuth2
                      config:
                        token_endpoint:
                          cluster: oauth
                          uri: "https://mykeycloak.mycluster/auth/realms/myrealm/protocol/openid-connect/token"
                          timeout: 5s
                        authorization_endpoint: "https://mykeycloak.mycluster/auth/realms/myrealm/protocol/openid-connect/auth"
                        redirect_uri: "http://%REQ(:authority)%/oauth2/callback"
                        redirect_path_matcher:
                          path:
                          exact: /oauth2/callback
                        signout_path:
                          path:
                            exact: /oauth2/signout
                        forward_bearer_token: true
                        credentials:
                          client_id: "envoy-redacted"
                          token_secret:
                            name: token
                            generic_secret:
                              secret:
                                inline_string: "redacted"
                          hmac_secret:
                            name: hmac
                            generic_secret:
                              secret:
                                inline_string: "redacted"
                        auth_scopes:
                          - openid
                  - name: envoy.filters.http.jwt_authn
                    typed_config:
                      "@type": type.googleapis.com/envoy.extensions.filters.http.jwt_authn.v3.JwtAuthentication
                      providers:
                        provider1:
                          remote_jwks:
                            http_uri:
                              uri: "https://mykeycloak.mycluster/auth/realms/myrealm/protocol/openid-connect/certs"
                              cluster: oauth
                              timeout: 5s
                            cache_duration: 600s
                      rules:
                        - match:
                            prefix: /
                          requires:
                            provider_name: provider1
                  - name: envoy.filters.http.router
                    typed_config: {}
    - "@type": type.googleapis.com/envoy.config.cluster.v3.Cluster
      name: oauth
      connect_timeout: 2s
      type: STRICT_DNS
      lb_policy: ROUND_ROBIN
      load_assignment:
       cluster_name: oauth
       endpoints:
         - lb_endpoints:
           - endpoint:
               address:
                 socket_address:
                   address: mykeycloak.mycluster
                   port_value: 443
      transport_socket:
        name: envoy.transport_sockets.tls
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
          sni: mykeycloak.mycluster

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by gsharma-jiggzy** [comment]
Just poking around in the code looks like this might be the area to address for a fix. Looks like `github.com/cilium/proxy/go/envoy/extensions/http/oauth/v3` is not added to the import here.
https://github.com/cilium/cilium/blob/68781336108faf86796699c6d4b6c8f5a2a3ca0e/pkg/envoy/ciliumenvoyconfig.go#L30

**Comment by jbgcarnes** [comment]
Is it feasible to get it added? I am very curious to have this added as well. (Looking to do the same thing, Envoy filter to require oidc auth)

**Comment by jwsy** [comment]
Bump

**Comment by cotocisternas** [comment]
Seams to be related to this [cilium/proxy #95](https://github.com/cilium/proxy/pull/95)

**Comment by gsharma-jiggzy** [comment]
some more information, I did just turn on envoy's daemon set and applied the clusterwide config. I did not get an error but the config did not trigger an auth.

**Comment by sayboras** [comment]
The decoding issue should be fixed as part of https://github.com/cilium/cilium/pull/28272.

> I did not get an error but the config did not trigger an auth.

I don't think it's related to Cilium provided that the filter is configured correctly. Can you share the config ?

**Comment by jwsy** [comment]
Unstale pls thx 🙏

**Comment by sayboras** [comment]
The decoding issue should be fixed as part of #28272, we can track any other potential issue separately.
